### PR TITLE
Add dynamic iframe config

### PR DESCRIPTION
**Pull request by cea2aj**

### CODE REVIEW

File: static/js/iframe-common.js [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
+ */
+export function generateIFrame(domain, hitchhikerConfigManager) {
   var isLocalHost = window.location.host.split(':')[0] === 'localhost';
   var containerEl = document.querySelector('#answers-container');
   var iframe = document.createElement('iframe');
   var pathToIndex = containerEl.dataset.path;
   iframe.allow = 'geolocation';
 
   domain = domain || '';
-  queryParam = queryParam || 'query';
```
Comment (cea2aj) [inline static/js/iframe-common.js:11]: I removed these because I could not find anywhere that they were being used

File: layouts/html.hbs [hunk 2]
```diff
@@ -126,6 +126,10 @@
       window.answersJWT = new Promise(resolve => {
         window.initAnswersJWT = resolve;
       });
+      const hhConfigMetadata = {
+        lastReceived: -1,
```
Comment (tmeyer2115) [inline layouts/html.hbs:130]: What would you think about adding a `lastUpdatedTimestamp` or something like that to the `ConfigManager` itself? We could then use that in the comparison below.

File: layouts/html.hbs [hunk 3]
```diff
@@ -136,8 +140,20 @@
...
+          const messageContainsNewHHConfig = message.hitchhikerConfig
+             && message.timestamp >= hhConfigMetadata.lastReceived;
+          if (messageContainsNewHHConfig) {
+            window.HitchhikerConfig = {
+              ...window.HitchhikerConfig,
+              ...message.hitchhikerConfig
+            };
+            hhConfigMetadata.lastReceived = message.timestamp;
+          }
+          if (message.hitchhikerConfig?.token && !hhConfigMetadata.hasInitializedWithJWT) {
```
Comment (tmeyer2115) [inline layouts/html.hbs:154]: Since Answers can't be initialized twice, I'm wondering if it would be beneficial to have a more general variable than `hasInitializedWithJWT`. Something like `hasAnswersInitialized`. We could then use that here and in other situations potentially.

File: static/js/hitchhikerconfigmanager.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  getObject () {
+    return { ...this._hitchhikerConfig };
+  }
+
+  set (key, value) {
+    this._validate(key, value);
+    this._hitchhikerConfig[key] = value;
+  }
+
+  sendConfigToIframe() {
```
Comment (tmeyer2115) [inline static/js/hitchhikerconfigmanager.js:20]: I think it might be a bit cleaner if we leave this logic to `iframe-common`. Without this method, we could just make this class `HitchhikerConfig.js`. It's purely a run-time data store. Consumers of the data store can add additional behaviors, such as sending the info to an iFrame via a message. In theory, someone could use this thing in the non-iFrame integration.

File: static/js/iframe-common.js (hunk 1 above)
Comment (oshi97) [reply static/js/iframe-common.js:11]: untangle one 🍝  a day to keep the smell away

File: static/js/utils/getHHConfigAttributes.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * Returns an object which contains the data-hhconfig attributes on the element specified by the selector
+ * 
+ * @example
+ * If the attributes on the element contain 'data-hhconfig-link-target="_top"', the returned
+ * object will be { link-target: "_top" }
+ *
+ * @param {string} selector
+ * @returns {Object}
+ */
+export default function getHHConfigAttributes (selector) {
```
Comment (oshi97) [inline static/js/utils/getHHConfigAttributes.js:11]: can we get some unit tests for this function? would make it easier to visualize as well

File: static/js/utils/getHHConfigAttributes.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  const element = document.querySelector(selector);
+  if (!element) {
+    return {};
+  }
+  const attributes = element.attributes;
```
Comment (oshi97) [inline static/js/utils/getHHConfigAttributes.js:16]: I think it'd be simpler to access the data-* attributes directly (`element.dataset`)

File: static/js/iframe-prod.js [hunk 7]
```diff
@@ -1,5 +1,9 @@
 import { generateIFrame } from './iframe-common';
 import InjectedData from './models/InjectedData';
+import { HitchhikerConfigManager, exposeOnWindow } from './hitchhikerconfigmanager';
+
+const hitchhikerConfigManager = new HitchhikerConfigManager();
+exposeOnWindow(hitchhikerConfigManager);
```
Comment (oshi97) [inline static/js/iframe-prod.js:6]: do the static iframe scripts need hitchhikerConfigManager?

File: test-site/public/iframe_test.html [hunk 8]
```diff
@@ -1 +1 @@
-<html><head><meta name="viewport" content="initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=no,width=device-width"></head><body><div id="answers-container"></div><script src="iframe.js"></script></body></html>
\ No newline at end of file
+<html><head><meta name="viewport" content="initial-scale=1,maximum-scale=1,minimum-scale=1,user-scalable=no,width=device-width"></head><body><div id="answers-container" data-hhconfig-test="test-data"></div><script src="iframe.js"></script></body></html>
```
Comment (oshi97) [inline test-site/public/iframe_test.html:1]: are we committing this file just for easier manual testing in the future?

File: static/js/utils/getInitAnswersFrame.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Returns a function which generates an iframed Answers experience.
+ * If the function is not called by the timeout, a warning is logged.
+ * 
+ * @param {string} domain 
+ * @param {HitchhikerConfigManager} hitchhikerConfigManager 
+ * @returns {Function}
+ */
+export default function getInitAnswersFrame (domain, hitchhikerConfigManager) {
```
Comment (oshi97) [inline static/js/utils/getInitAnswersFrame.js:11]: would something like generateInitAnswersFrame be better? We're not getting a preexisting function but generating a new one

File: static/js/hitchhikerconfigmanager.js (hunk 4 above)
Comment (cea2aj) [reply static/js/hitchhikerconfigmanager.js:20]: sure, that sounds good

File: layouts/html.hbs (hunk 2 above)
Comment (cea2aj) [reply layouts/html.hbs:130]: Sure, and include that data within the hitchhikerConfig data which is sent (as opposed to having it at the top level of the message)?

File: static/js/hitchhikerconfigmanager.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    }
+  }
+}
+
+/**
+ * Puts a setter on the window which allows the hitchhiker config to be updated
+ *  
+ * @param {HitchhikerConfigManager} hitchhikerConfigManager 
+ */
+export function exposeOnWindow (hitchhikerConfigManager) {
```
[the remark this answers is not in the document]
Comment (cea2aj) [reply static/js/hitchhikerconfigmanager.js:50]: ya, that makes sense

File: static/js/iframe-prod.js (hunk 7 above)
Comment (cea2aj) [reply static/js/iframe-prod.js:6]: ya, I will check with product, but having it here allows the someone with a regular iframe experience to update the config from the parent, even it it doesn't happen during the init

File: test-site/public/iframe_test.html (hunk 8 above)
Comment (cea2aj) [reply test-site/public/iframe_test.html:1]: Ya, I feel that it's helpful to have an example of using the data attribute

File: static/js/utils/getInitAnswersFrame.js (hunk 9 above)
Comment (cea2aj) [reply static/js/utils/getInitAnswersFrame.js:11]: ya, that works

File: static/js/utils/getHHConfigAttributes.js (hunk 6 above)
Comment (cea2aj) [reply static/js/utils/getHHConfigAttributes.js:16]: I was hesitant to use that due to the automatic conversion to camelcase which I thought might make complicate things. I'll get product's input on it

Comment (oshi97) [reply static/js/utils/getHHConfigAttributes.js:16]: ah that's fair, feel free to ack this

File: layouts/html.hbs (hunk 3 above)
Comment (cea2aj) [reply layouts/html.hbs]: updated!

File: static/js/utils/getHHConfigAttributes.js (hunk 5 above)
Comment (cea2aj) [reply static/js/utils/getHHConfigAttributes.js]: switched to data attributes so unit tests are no longer necessary

File: layouts/html.hbs [hunk 11]
```diff
@@ -123,9 +118,36 @@
...
+            updateRuntimeConfig(config);
+            clearTimeout(runtimeConfigNotProvidedTimeout);
+            initAnswers();
+          };
+        {{else}}
+          initAnswers();
+        {{/if}}
       });
+      const answersMetadata = {
+        hasInitialized: false
```
Comment (oshi97) [inline layouts/html.hbs:149]: would it be out of scope to refactor the window.iframeResizer logic into some static js file? maybe we can also do something about passing RuntimeConfig over the window?

Comment (cea2aj) [reply layouts/html.hbs:149]: Can you give a little more info on how that would work?

**Review by tmeyer2115 — COMMENTED** [review]
I think it's great that we have an actual domain model for `RuntimeConfig`. It's a first-class thing in the Theme, so it definitely deserves it. I'm wondering if we should add some other class(es) around managing it. I'm a bit wary of adding even more complexity to the `layouts/html.hbs` partial. There's a lot going on in there already. 

I think anything we can do to factor some of the complexity of run-time configuration into other files/classes would be good. I know this is kind of hand-wavy. I will try and come up with some more concrete ideas. Maybe we have some kind of `RuntimeConfigDirector` that does de-duping of updates and handles message passing to frames. 

In general, if we have complex JS, I think we should try and put as much of it as we can into static. By putting things in `static/js` we can manage complexity better with separate classes, etc. Complicated JS in partials is difficult to maintain. Historically, it's what we've done, but we didn't always have time to do things the "best way".

**Comment by cea2aj** [comment]
> I think it's great that we have an actual domain model for `RuntimeConfig`. It's a first-class thing in the Theme, so it definitely deserves it. I'm wondering if we should add some other class(es) around managing it. I'm a bit wary of adding even more complexity to the `layouts/html.hbs` partial. There's a lot going on in there already.
> 
> I think anything we can do to factor some of the complexity of run-time configuration into other files/classes would be good. I know this is kind of hand-wavy. I will try and come up with some more concrete ideas. Maybe we have some kind of `RuntimeConfigDirector` that does de-duping of updates and handles message passing to frames.
> 
> In general, if we have complex JS, I think we should try and put as much of it as we can into static. By putting things in `static/js` we can manage complexity better with separate classes, etc. Complicated JS in partials is difficult to maintain. Historically, it's what we've done, but we didn't always have time to do things the "best way".

That makes sense. I will experiment with putting this logic in HitchhikerJS

File: layouts/html.hbs (hunk 11 above)
Comment (oshi97) [reply layouts/html.hbs:149]: sorry I didn't mean to make it sound like it was easy, and didn't have really concrete plans in mind. I was imagining moving this DOMContentLoaded listener into a static file (the on-document-loaded partial would probably stay here. we can just make them separate listeners). For the RuntimeConfig, instead of getting it from the window in core.hbs, maybe we could have a HitchhikerJS.getRuntimeConfig() ? I also thought we might want to refactor window.iFrameResizer into it's own file, but looking at it again that can probably wait.

Comment (cea2aj) [reply layouts/html.hbs]: I updated the new code to be inside static files rather than in the handlebars. I will leave the refactoring of the window.iFrameResizer to another PR since that is out of scope here. Changing it to HitchhikerJS.getRuntimeConfig() would also change the public interface for accessing the RuntimeConfig. Maybe I should check with product on that? Personally I think that accessing 'window.RuntimeConfig.linkTarget` is a little simpler than 'HitchhikerJS.getRuntimeConfig().linkTarget', since Hitchhikers will be referencing it from inside the cards.

File: layouts/html.hbs [hunk 12]
```diff
@@ -136,13 +134,19 @@
...
           }
-          window.isOverlay && window.Overlay.onMessage(message);
         }
       };
+      {{#if global_config.initializeManually}}
+        const initializeManually = true;
+      {{else}}
+        const initializeManually = false;
+      {{/if}}
+      const experienceInitializer = new HitchhikerJS.ExperienceInitializer(initializeManually).setup();
```
Comment (tmeyer2115) [inline layouts/html.hbs:149]: Do we only need an `experienceInitializer` in the `initializeManually` case or no?

File: static/js/experience-initializer.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+      })
+    }
+  }
+
+  /**
+   * Executes the provided function on DOM load
+   * @param {Function} cb 
+   */
+  _fireOnDomLoad (cb) {
+    if  (document.readyState == 'loading') {
```
Comment (oshi97) [inline static/js/experience-initializer.js:31]: nit extra space

File: static/js/experience-initializer.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    this._initializeManually = initializeManually;
+    this._hasAnswersInitialized = false;
+  }
+
+  /**
+   * Initializes on DOM load, or puts an init function on the window if initializeManually is true
+   */
+  setup () {
+    if (this._initializeManually) {
+      window.initAnswersExperience = this._generateInitFunction();
```
Comment (tmeyer2115) [inline static/js/experience-initializer.js:17]: Does the iFrame integration end up invoking this as well?

**Review by oshi97 — COMMENTED** [review]
besides nit lgtm!

File: static/js/iframe-prod.js [hunk 15]
```diff
@@ -1,5 +1,15 @@
...
+const runtimeConfig = new RuntimeConfig();
+window.setRuntimeConfig = (key, value) => {
+  runtimeConfig.set(key, value);
+  sendToIframe({ runtimeConfig: runtimeConfig.getObject() });
+}
+
+const prodDomain = new InjectedData().getProdDomain();
+generateIFrame(prodDomain, runtimeConfig);
+
+window.initAnswersExperienceFrame = generateInitAnswersFrame(runtimeConfig);
```
Comment (tmeyer2115) [inline static/js/iframe-prod.js:15]: Here and in the other iFrame script, are we waiting for them to set the `runtimeConfig` before calling `initAnswers`? If so, it's a little hard to follow how that's happening. For my own edification, could you explain that flow?

**Review by tmeyer2115 — COMMENTED** [review]
This is definitely much more stream-lined! Having these additional classes is really helpful. It allows us to compartmentalize the complexity better.

File: layouts/html.hbs (hunk 12 above)
Comment (cea2aj) [reply layouts/html.hbs:149]: Currently it's used for both automatic and manual initialization. I could re-work it so that the automatic initialization occurs in layouts/html.hbs, and the experience initializer is just for initializeManually. (I would also probably update the naming). That might be a better separation of concerns

File: static/js/experience-initializer.js (hunk 14 above)
Comment (cea2aj) [reply static/js/experience-initializer.js:17]: Yes, if the RuntimeConfigReceiver gets a message from the parent with initAnswersFrame set to true, then it will call the same initAnswersExperience function

File: static/js/iframe-prod.js (hunk 15 above)
Comment (cea2aj) [reply static/js/iframe-prod.js:15]: Ah, after refactoring the 'generateInitAnswersFrame' naming is no longer entirely accurate. I will update it. What happens is that the iframe is always generated immediately. If there were any data attributes set, those will be included in the runtime config during initialization. Next, the `initAnswersExperienceFrame` function is put on the window which updates the runtime config and sends the message to the iframe that the Answers experience should initialize

File: layouts/html.hbs (hunk 12 above)
Comment (tmeyer2115) [reply layouts/html.hbs:149]: Got it, I'm fine either way!

**Review by tmeyer2115 — APPROVED** [review]

File: static/js/manual-initializer.js [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        'An initialization function has not been called within 5 seconds of page load, and "initializeManually" is set to true.\n' +
+        'Load the experience by calling initAnswersExperience() or initAnswersExperienceFrame() for iframe integrations.'
+      );
+    }, 5000);
+    return runtimeConfig => {
+      this._fireOnDomLoad(() => {
+        updateRuntimeConfig(runtimeConfig);
+        clearTimeout(runtimeConfigNotProvidedTimeout);
+        if (!this._hasAnswersInitialized) {
+          initAnswers();
```
Comment (oshi97) [inline static/js/manual-initializer.js:34]: could we pass in initAnswers into the constructor?

Comment (cea2aj) [reply static/js/manual-initializer.js:34]: sure, that's a nice way of reducing the reliance on globals

Comment (cea2aj) [reply static/js/manual-initializer.js]: updated to pass in the init function

**Review by oshi97 — APPROVED** [review]
lgtm! 🐹 🐹 🐹  happy about all the progress
